staging/src/k8s.io/apiserver/pkg/cel/library/ip.go

```diff
@@ -0,0 +1,307 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"fmt"
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// IPs provides a CEL function library extension of IP address parsing functions.
+//
+// ip
+//
+// Converts a string to an IP address or results in an error if the string is not a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+// Leading zeros in IPv4 address octets are not allowed.
+//
+//	ip(<string>) <IPAddr>
+//
+// Examples:
+//
+//	ip('127.0.0.1') // returns an IPv4 address
+//	ip('::1') // returns an IPv6 address
+//	ip('127.0.0.256') // error
+//	ip(':::1') // error
+//
+// isIP
+//
+// Returns true if a string is a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+// Leading zeros in IPv4 address octets are not allowed.
+//
+//	isIP(<string>) <bool>
+//
+// Examples:
+//
+//	isIP('127.0.0.1') // returns true
+//	isIP('::1') // returns true
+//	isIP('127.0.0.256') // returns false
+//	isIP(':::1') // returns false
+//
+// family / isCanonical / isInCIDR / isUnspecified / isLoopback / isLinkLocalMulticast / isLinkLocalUnicast
+//
+// Returns true if a string is a valid IPv4 address, IPv6 address, or is in its canonical format.
+//
+// - family: returns the IP addresses' family (IPv4 or IPv6) as an integer, either '4' or '6'.
+//
+// - isCanonical: returns true when the IP Address is in its canonical format. There is exactly
+// one canonical form for every IP address, so fields containing IPs in canonical
+// form can just be treated as strings when checking for equality or uniqueness.
+//
+// - isInCIDR: returns true if the IP address is contained in the CIDR.
```

Comment:
No worries about the nit picking, I understand that once set these can't be changed so we need to get this right!

I've been stuck in Go for so long, that I forgot overloads are a thing. I didn't like the `contains` idea originally because I was thinking we would need two contains methods , one for IP, one for CIDR. But it turns out CEL supports overloading of methods so, we can just create one `contains` with two implementations, and have the user facing side accept either an IP or a CIDR.

Would you be happy with that UX?

Reply:
Is overloading "normal" in CEL?  I think we should follow as many normal conventions as possible :)

I am fine with cidr.contains() or cidr.containsIP() + cidr.containsCIDR or (last pref :) ip.isInCIDR().  You probably know CEL norms better than me.

Reply:
I can see overloads in the authz, quantity and regex libraries. Each has at least one method that has the option to take different arguments of different types. So it looks like there is precedence.

Given that, we could make this work with three overloads, `contains(<string>)`, `contains(ip(<string>))` and `contains(cidr(<string>))` where the raw string case would have to check if the string is an IP or CIDR first before delegating to the appropriate contains.

I'd probably want @cici37 or @jpbetz to sign off on the overloading like this

Reply:
CEL overloading is quite comprehensive and heavily relied on.

I'm supportive of using overloads where the function's declared purpose is the same across all overloads.  I would try to avoid overloads where the documentation will have to define distinct semantics for each possible overload.

In this case, I think I'm in favor of a `containsIP()` function overloaded to take either a string or an IP, and a separate `containsCIDR()` function overloaded to take either a string or a CIDR.

Reply:
cc @TristonianJones for any additional thoughts

Reply:
Meeting the user where they are is great for helping users be successful in writing an expression that 'just works'; However, this can lead to suboptimal performance when the overloads handle a lot of things which would be better to do once, e.g. parse the ip and then perform operations on it.

If you force the user to convert the string to an ip, then you'll get code which is more exact, but not necessarily more clear in its intent.

Since you have cost tracking to encourage optimization, then it seems reasonable to weight the convenience overloads as more costly, but keep them to increase the odds that the expression 'just works'